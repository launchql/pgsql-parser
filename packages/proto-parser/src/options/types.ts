export type PgProtoParserOptions = PgProtoStoreOptions & {
    parser?: ParseProtoOptions;
};

/**
 * Configuration options for PgProtoStore.
 */
export interface PgProtoStoreOptions {
    // The directory where the generated files will be saved.
    outDir?: string;

    // Types and Enums to skip during processing
    exclude?: string[],

    // Options related to utility functions.
    utils?: {
        enums?: {
            // Whether to include TypeScript utils for enums
            enabled?: boolean;
            // Enums helpers filename
            filename?: string;
            // Whether to generate separate unidirectional functions
            unidirectional?: boolean;
            // Filename for enum to int function
            toIntFilename?: string;
            // Filename for enum to string function
            toStringFilename?: string;
            // Output format: 'switchStatements' or 'nestedObjects'
            outputFormat?: 'switchStatements' | 'nestedObjects';
        }

        // AST helper functions.
        astHelpers?: {
            // Whether to include TypeScript AST builders
            enabled?: boolean;
            // Source file for AST helper types.
            typesSource?: string;
            // Whether to inline nested-obj
            inlineNestedObj?: boolean;
            // if inlined, filename
            nestedObjFile?: string;
            // ASTs helpers filename
            filename?: string;
        }
    };

    // Options related to TypeScript types.
    types?: {
        // Whether to generate TypeScript interfaces for protobuf messages.
        enabled?: boolean;
        // Types filename
        filename?: string;
        // Whether fields in TypeScript interfaces should be optional.
        optionalFields?: boolean;
        // Enums source specifier
        enumsSource?: string;
        // Node Object Keys, e.g. { ParseResult: ParseResult, ... }
        // meant for simpler user cases
        // if a field in an object uses Node, it's a node type
        // if a field in an object references a type, e.g. RangeVar, it's flat
        wrappedNodeTypeExport?: boolean;
    };

    // Options related to runtime schema generation.
    runtimeSchema?: {
        // Whether to generate runtime schema for AST nodes
        enabled?: boolean;
        filename?: string;
        format?: 'json' | 'typescript';
    };

    // Options related to enumeration handling.
    enums?: {
        // Whether to include TypeScript enums.
        enabled?: boolean;
        // Enums filename
        filename?: string;
        // Whether enums in TypeScript are a union type
        enumsAsTypeUnion?: boolean;
        // Options for enum mappings output.
        enumMap?: {
            // Whether to generate enum mapping files.
            enabled?: boolean;
            // Output format: 'json' for plain JSON files, 'ts' for TypeScript exports
            format?: 'json' | 'ts';
            toIntOutFile?: string;
            toStrOutFile?: string;
        };
        // Whether to remove the initial `UNDEFINED` enum entry and adjust subsequent values.
        removeUndefinedAt0?: boolean;
    };

    // Whether to include a header in generated TypeScript files.
    includeHeader?: boolean;
}


export interface ParseProtoOptions {
    keepCase?: boolean;
    alternateCommentMode?: boolean;
    preferTrailingComment?: boolean;
}
