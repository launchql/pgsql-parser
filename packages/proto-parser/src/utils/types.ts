export const isPrimitiveType = (type: string) => {
    switch (type) {
      case 'string':
      case 'double':
      case 'float':
      case 'int32':
      case 'uint32':
      case 'sint32':
      case 'fixed32':
      case 'sfixed32':
      case 'int64':
      case 'uint64':
      case 'sint64':
      case 'fixed64':
      case 'sfixed64':
      case 'bytes':
      case 'bool':
        return true;
      default:
        return false;
    };
  };