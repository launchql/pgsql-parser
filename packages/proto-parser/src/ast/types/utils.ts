import * as t from '@babel/types';
import { isPrimitiveType } from '../../utils';

export const getTSType = (type: string = 'any') => {
  switch (type) {
    case 'string':
      return t.tsStringKeyword();
    case 'double':
    case 'float':
    case 'int32':
    case 'uint32':
    case 'sint32':
    case 'fixed32':
    case 'sfixed32':
      return t.tsNumberKeyword();
    case 'int64':
    case 'uint64':
    case 'sint64':
    case 'fixed64':
    case 'sfixed64':
      return t.tsBigIntKeyword()
    case 'bytes':
      return t.tsTypeReference(t.identifier('Uint8Array'));
    case 'bool':
      return t.tsBooleanKeyword();
    default:
      // For custom types, reference them directly
      return t.tsTypeReference(t.identifier(type));

  };
};

export const resolveTypeName = (type: string) => {
  if (isPrimitiveType(type)) {
    return getTSType(type);
  }
  return t.tsTypeReference(t.identifier(type));
}
