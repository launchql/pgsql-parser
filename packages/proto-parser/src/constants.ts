export const NODE_TYPE = 'Node';
