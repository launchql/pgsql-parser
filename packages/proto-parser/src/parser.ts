import { parse } from '@launchql/protobufjs';
import { readFileSync } from 'fs';
import { ProtoStore } from './store';
import {
  getOptionsWithDefaults,
  PgProtoParserOptions
} from './options';

export class PgProtoParser {
  inFile: string;
  options: PgProtoParserOptions;

  constructor(inFile: string, options?: PgProtoParserOptions) {
    this.inFile = inFile;
    this.options = getOptionsWithDefaults(options);
  }

  private readProtoFile(): string {
    return readFileSync(this.inFile, 'utf-8');
  }

  private parseProto(content: string): any {
    return parse(content, this.options.parser);
  }

  public write(): void {
    const protoContent = this.readProtoFile();
    const ast = this.parseProto(protoContent);
    const store = new ProtoStore(ast.root, this.options);
    store.write();
  }
}
