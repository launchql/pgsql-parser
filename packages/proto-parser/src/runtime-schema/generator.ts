import { Type, Field, Enum, Namespace, ReflectionObject } from '@launchql/protobufjs';
import { NodeSpec, FieldSpec } from './types';

export class RuntimeSchemaGenerator {
  private root: Namespace;
  private nodeTypes: Set<string> = new Set();

  constructor(root: Namespace) {
    this.root = root;
    this.extractNodeTypes();
  }

  private extractNodeTypes(): void {
    const nodeType = this.root.lookupType('Node');
    if (nodeType && nodeType.oneofs && nodeType.oneofs.node) {
      const oneof = nodeType.oneofs.node;
      for (const fieldName of oneof.fieldsArray.map(f => f.name)) {
        const field = nodeType.fields[fieldName];
        if (field && field.type) {
          this.nodeTypes.add(field.type);
        }
      }
    }
  }

  public generateNodeSpecs(): NodeSpec[] {
    const nodeSpecs: NodeSpec[] = [];
    
    const pgQueryNamespace = this.root.nested?.pg_query;
    if (pgQueryNamespace && pgQueryNamespace instanceof Namespace && pgQueryNamespace.nestedArray) {
      pgQueryNamespace.nestedArray.forEach((nested) => {
        if (nested instanceof Type && nested.name !== 'Node') {
          const nodeSpec = this.createNodeSpec(nested);
          if (nodeSpec) {
            nodeSpecs.push(nodeSpec);
          }
        }
      });
    }

    return nodeSpecs.sort((a, b) => a.name.localeCompare(b.name));
  }

  private createNodeSpec(type: Type): NodeSpec | null {
    const fields: FieldSpec[] = [];
    
    Object.values(type.fields).forEach((field) => {
      const fieldSpec = this.createFieldSpec(field);
      if (fieldSpec) {
        fields.push(fieldSpec);
      }
    });

    return {
      name: type.name,
      fields: fields.sort((a, b) => a.name.localeCompare(b.name))
    };
  }

  private createFieldSpec(field: Field): FieldSpec | null {
    const fieldName = field.name;
    const isArray = field.repeated || false;
    const optional = !field.required;
    const fieldType = field.type;

    return {
      name: fieldName,
      type: fieldType,
      isArray,
      optional
    };
  }

  public getNodeTypes(): string[] {
    return Array.from(this.nodeTypes).sort();
  }

  public getNodeTypesCount(): number {
    return this.nodeTypes.size;
  }
}
