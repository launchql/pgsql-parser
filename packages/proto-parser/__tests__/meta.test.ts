import * as t from '../test-utils/meta';
import { SelectStmt } from '@pgsql/types';
import { generateTsAstCodeFromPgAst, generateTsAstCodeFromPgAstWithSchema } from '../src/utils'
import { runtimeSchema } from '../test-utils/meta/runtime-schema';
import generate from '@babel/generator';

it('AST to AST to create AST â€”Â meta ðŸ¤¯', () => {
    const selectStmt = t.nodes.selectStmt({
        targetList: [
          t.nodes.resTarget({
            val: t.nodes.columnRef({
              fields: [t.nodes.aStar()]
            })
          })
        ],
        fromClause: [
          t.nodes.rangeVar({
            relname: 'some_amazing_table',
            inh: true,
            relpersistence: 'p'
          })
        ],
        limitOption: 'LIMIT_OPTION_DEFAULT',
        op: 'SETOP_NONE'
      });

      expect(selectStmt).toMatchSnapshot();
      
      const astForAst = generateTsAstCodeFromPgAst(selectStmt);
      expect(generate(astForAst).code).toMatchSnapshot();
});

it('Complex AST â€” Advanced SQL with CTEs, Window Functions, Joins, and Subqueries', () => {
    const complexSelectStmt: { SelectStmt: SelectStmt } = t.nodes.selectStmt({
      // WITH clause for CTEs
      withClause: t.ast.withClause({
        ctes: [
          // First CTE: Sales summary
          t.nodes.commonTableExpr({
            ctename: 'sales_summary',
            ctequery: t.nodes.selectStmt({
              targetList: [
                t.nodes.resTarget({
                  name: 'customer_id',
                  val: t.nodes.columnRef({
                    fields: [t.nodes.string({ sval: 'customer_id' })]
                  })
                }),
                t.nodes.resTarget({
                  name: 'total_sales',
                  val: t.nodes.funcCall({
                    funcname: [t.nodes.string({ sval: 'sum' })],
                    args: [
                      t.nodes.columnRef({
                        fields: [t.nodes.string({ sval: 'amount' })]
                      })
                    ]
                  })
                }),
                t.nodes.resTarget({
                  name: 'avg_order_value',
                  val: t.nodes.funcCall({
                    funcname: [t.nodes.string({ sval: 'avg' })],
                    args: [
                      t.nodes.columnRef({
                        fields: [t.nodes.string({ sval: 'amount' })]
                      })
                    ]
                  })
                })
              ],
              fromClause: [
                t.nodes.rangeVar({
                  relname: 'orders',
                  inh: true,
                  relpersistence: 'p'
                })
              ],
              whereClause: t.nodes.aExpr({
                kind: 'AEXPR_OP',
                name: [t.nodes.string({ sval: '>=' })],
                lexpr: t.nodes.columnRef({
                  fields: [t.nodes.string({ sval: 'order_date' })]
                }),
                rexpr: t.nodes.aConst({
                  sval: t.ast.string({ sval: '2023-01-01' })
                })
              }),
              groupClause: [
                t.nodes.columnRef({
                  fields: [t.nodes.string({ sval: 'customer_id' })]
                })
              ],
              limitOption: 'LIMIT_OPTION_DEFAULT',
              op: 'SETOP_NONE'
            })
          }),
          t.nodes.commonTableExpr({
            ctename: 'customer_rankings',
            ctequery: t.nodes.selectStmt({
              targetList: [
                t.nodes.resTarget({
                  name: 'customer_id',
                  val: t.nodes.columnRef({
                    fields: [t.nodes.string({ sval: 'customer_id' })]
                  })
                }),
                t.nodes.resTarget({
                  name: 'total_sales',
                  val: t.nodes.columnRef({
                    fields: [t.nodes.string({ sval: 'total_sales' })]
                  })
                }),
                t.nodes.resTarget({
                  name: 'sales_rank',
                  val: t.nodes.windowFunc({
                    winfnoid: 3133, // ROW_NUMBER function OID
                    wintype: 20, // INT8 type
                    args: [],
                    winref: 1,
                    winstar: false,
                    winagg: false
                  })
                }),
                t.nodes.resTarget({
                  name: 'sales_percentile',
                  val: t.nodes.windowFunc({
                    winfnoid: 3974, // PERCENT_RANK function OID  
                    wintype: 701, // FLOAT8 type
                    args: [],
                    winref: 2,
                    winstar: false,
                    winagg: false
                  })
                })
              ],
              fromClause: [
                t.nodes.rangeVar({
                  relname: 'sales_summary',
                  inh: true,
                  relpersistence: 'p'
                })
              ],
              windowClause: [
                t.nodes.windowDef({
                  name: 'sales_window',
                  orderClause: [
                    t.nodes.sortBy({
                      node: t.nodes.columnRef({
                        fields: [t.nodes.string({ sval: 'total_sales' })]
                      }),
                      sortby_dir: 'SORTBY_DESC',
                      sortby_nulls: 'SORTBY_NULLS_DEFAULT'
                    })
                  ]
                })
              ],
              limitOption: 'LIMIT_OPTION_DEFAULT',
              op: 'SETOP_NONE'
            })
          })
        ],
        recursive: false
      }),
      
      // Main SELECT target list with complex expressions
      targetList: [
        t.nodes.resTarget({
          name: 'customer_name',
          val: t.nodes.columnRef({
            fields: [t.nodes.string({ sval: 'c' }), t.nodes.string({ sval: 'name' })]
          })
        }),
        t.nodes.resTarget({
          name: 'total_sales',
          val: t.nodes.columnRef({
            fields: [t.nodes.string({ sval: 'cr' }), t.nodes.string({ sval: 'total_sales' })]
          })
        }),
        t.nodes.resTarget({
          name: 'sales_rank',
          val: t.nodes.columnRef({
            fields: [t.nodes.string({ sval: 'cr' }), t.nodes.string({ sval: 'sales_rank' })]
          })
        }),
        t.nodes.resTarget({
          name: 'customer_tier',
          val: t.nodes.caseExpr({
            args: [
              t.nodes.caseWhen({
                expr: t.nodes.aExpr({
                  kind: 'AEXPR_OP',
                  name: [t.nodes.string({ sval: '<=' })],
                  lexpr: t.nodes.columnRef({
                    fields: [t.nodes.string({ sval: 'cr' }), t.nodes.string({ sval: 'sales_rank' })]
                  }),
                  rexpr: t.nodes.aConst({
                    ival: t.ast.integer({ ival: 10 })
                  })
                }),
                result: t.nodes.aConst({
                  sval: t.ast.string({ sval: 'Premium' })
                })
              }),
              t.nodes.caseWhen({
                expr: t.nodes.aExpr({
                  kind: 'AEXPR_OP',
                  name: [t.nodes.string({ sval: '<=' })],
                  lexpr: t.nodes.columnRef({
                    fields: [t.nodes.string({ sval: 'cr' }), t.nodes.string({ sval: 'sales_rank' })]
                  }),
                  rexpr: t.nodes.aConst({
                    ival: t.ast.integer({ ival: 50 })
                  })
                }),
                result: t.nodes.aConst({
                  sval: t.ast.string({ sval: 'Gold' })
                })
              }),
              t.nodes.caseWhen({
                expr: t.nodes.aExpr({
                  kind: 'AEXPR_OP',
                  name: [t.nodes.string({ sval: '<=' })],
                  lexpr: t.nodes.columnRef({
                    fields: [t.nodes.string({ sval: 'cr' }), t.nodes.string({ sval: 'sales_rank' })]
                  }),
                  rexpr: t.nodes.aConst({
                    ival: t.ast.integer({ ival: 100 })
                  })
                }),
                result: t.nodes.aConst({
                  sval: t.ast.string({ sval: 'Silver' })
                })
              })
            ],
            defresult: t.nodes.aConst({
              sval: t.ast.string({ sval: 'Bronze' })
            })
          })
        }),
        t.nodes.resTarget({
          name: 'recent_order_count',
          val: t.nodes.subLink({
            subLinkType: 'EXPR_SUBLINK',
            subselect: t.nodes.selectStmt({
              targetList: [
                t.nodes.resTarget({
                  val: t.nodes.funcCall({
                    funcname: [t.nodes.string({ sval: 'count' })],
                    args: [t.nodes.aStar()]
                  })
                })
              ],
              fromClause: [
                t.nodes.rangeVar({
                  relname: 'orders',
                  alias: t.ast.alias({ aliasname: 'o2' }),
                  inh: true,
                  relpersistence: 'p'
                })
              ],
              whereClause: t.nodes.boolExpr({
                boolop: 'AND_EXPR',
                args: [
                  t.nodes.aExpr({
                    kind: 'AEXPR_OP',
                    name: [t.nodes.string({ sval: '=' })],
                    lexpr: t.nodes.columnRef({
                      fields: [t.nodes.string({ sval: 'o2' }), t.nodes.string({ sval: 'customer_id' })]
                    }),
                    rexpr: t.nodes.columnRef({
                      fields: [t.nodes.string({ sval: 'c' }), t.nodes.string({ sval: 'id' })]
                    })
                  }),
                  t.nodes.aExpr({
                    kind: 'AEXPR_OP',
                    name: [t.nodes.string({ sval: '>=' })],
                    lexpr: t.nodes.columnRef({
                      fields: [t.nodes.string({ sval: 'o2' }), t.nodes.string({ sval: 'order_date' })]
                    }),
                    rexpr: t.nodes.funcCall({
                      funcname: [t.nodes.string({ sval: 'current_date' })],
                      args: []
                    })
                  })
                ]
              }),
              limitOption: 'LIMIT_OPTION_DEFAULT',
              op: 'SETOP_NONE'
            })
          })
        })
      ],
      
      // Complex FROM clause with joins
      fromClause: [
        t.nodes.joinExpr({
          jointype: 'JOIN_INNER',
          larg: t.nodes.rangeVar({
            relname: 'customers',
            alias: t.ast.alias({ aliasname: 'c' }),
            inh: true,
            relpersistence: 'p'
          }),
          rarg: t.nodes.rangeVar({
            relname: 'customer_rankings',
            alias: t.ast.alias({ aliasname: 'cr' }),
            inh: true,
            relpersistence: 'p'
          }),
          quals: t.nodes.aExpr({
            kind: 'AEXPR_OP',
            name: [t.nodes.string({ sval: '=' })],
            lexpr: t.nodes.columnRef({
              fields: [t.nodes.string({ sval: 'c' }), t.nodes.string({ sval: 'id' })]
            }),
            rexpr: t.nodes.columnRef({
              fields: [t.nodes.string({ sval: 'cr' }), t.nodes.string({ sval: 'customer_id' })]
            })
          })
        })
      ],
      
      // WHERE clause with complex conditions
      whereClause: t.nodes.boolExpr({
        boolop: 'AND_EXPR',
        args: [
          t.nodes.aExpr({
            kind: 'AEXPR_OP',
            name: [t.nodes.string({ sval: '>' })],
            lexpr: t.nodes.columnRef({
              fields: [t.nodes.string({ sval: 'cr' }), t.nodes.string({ sval: 'total_sales' })]
            }),
            rexpr: t.nodes.aConst({
              ival: t.ast.integer({ ival: 1000 })
            })
          }),
          t.nodes.aExpr({
            kind: 'AEXPR_OP',
            name: [t.nodes.string({ sval: 'IS NOT' })],
            lexpr: t.nodes.columnRef({
              fields: [t.nodes.string({ sval: 'c' }), t.nodes.string({ sval: 'status' })]
            }),
            rexpr: t.nodes.aConst({
              sval: t.ast.string({ sval: 'inactive' })
            })
          })
        ]
      }),
      
      // ORDER BY clause
      sortClause: [
        t.nodes.sortBy({
          node: t.nodes.columnRef({
            fields: [t.nodes.string({ sval: 'cr' }), t.nodes.string({ sval: 'sales_rank' })]
          }),
          sortby_dir: 'SORTBY_ASC',
          sortby_nulls: 'SORTBY_NULLS_DEFAULT'
        })
      ],
      
      // LIMIT clause
      limitCount: t.nodes.aConst({
        ival: t.ast.integer({ ival: 50 })
      }),
      
      limitOption: 'LIMIT_OPTION_COUNT',
      op: 'SETOP_NONE'
    });

    expect(complexSelectStmt).toMatchSnapshot();
    
    const astForComplexAst = generateTsAstCodeFromPgAst(complexSelectStmt);
    expect(generate(astForComplexAst).code).toMatchSnapshot();
});

it('Enhanced AST generation with runtime schema â€” wrapped vs unwrapped nodes', () => {
    const selectStmt = t.nodes.selectStmt({
        targetList: [
            t.nodes.resTarget({
                val: t.nodes.columnRef({
                    fields: [t.nodes.aStar()]
                })
            })
        ],
        fromClause: [
            t.nodes.rangeVar({
                relname: 'test_table',
                inh: true,
                relpersistence: 'p'
            })
        ],
        limitOption: 'LIMIT_OPTION_DEFAULT',
        op: 'SETOP_NONE'
    });

    expect(selectStmt).toMatchSnapshot();
    
    const enhancedAst = generateTsAstCodeFromPgAstWithSchema(selectStmt, runtimeSchema);
    const generatedCode = generate(enhancedAst).code;
    
    expect(generatedCode).toMatchSnapshot();
    
    expect(generatedCode).toContain('t.nodes.selectStmt');
    expect(generatedCode).toContain('t.nodes.resTarget');
});

it('Complex AST with runtime schema â€” mixed wrapped/unwrapped patterns', () => {
    const complexStmt = t.nodes.selectStmt({
        withClause: t.ast.withClause({
            ctes: [
                t.nodes.commonTableExpr({
                    ctename: 'test_cte',
                    ctequery: t.nodes.selectStmt({
                        targetList: [
                            t.nodes.resTarget({
                                val: t.nodes.columnRef({
                                    fields: [t.nodes.string({ sval: 'id' })]
                                })
                            })
                        ],
                        limitOption: 'LIMIT_OPTION_DEFAULT'
                    })
                })
            ],
            recursive: false
        }),
        targetList: [
            t.nodes.resTarget({
                val: t.nodes.columnRef({
                    fields: [t.nodes.aStar()]
                })
            })
        ],
        limitOption: 'LIMIT_OPTION_DEFAULT',
        op: 'SETOP_NONE'
    });

    const enhancedAst = generateTsAstCodeFromPgAstWithSchema(complexStmt, runtimeSchema);
    const generatedCode = generate(enhancedAst).code;
    
    expect(generatedCode).toMatchSnapshot();
    expect(generatedCode).toContain('t.ast.withClause');
    expect(generatedCode).toContain('t.nodes.selectStmt');
});
