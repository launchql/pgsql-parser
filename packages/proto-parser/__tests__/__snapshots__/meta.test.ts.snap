// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AST to AST to create AST â€”Â meta ðŸ¤¯ 1`] = `
"ast.selectStmt({
  targetList: [ast.resTarget({
    val: ast.columnRef({
      fields: [ast.aStar({})]
    })
  })],
  fromClause: [ast.rangeVar({
    relname: "some_amazing_table",
    inh: true,
    relpersistence: "p"
  })],
  limitOption: "LIMIT_OPTION_DEFAULT",
  op: "SETOP_NONE"
})"
`;

exports[`Complex AST with runtime schema â€” mixed wrapped/unwrapped patterns 1`] = `
"t.nodes.selectStmt({
  targetList: [t.nodes.resTarget({
    val: t.nodes.columnRef({
      fields: [t.nodes.aStar({})]
    })
  })],
  limitOption: "LIMIT_OPTION_DEFAULT",
  withClause: t.ast.withClause({
    ctes: [t.nodes.commonTableExpr({
      ctename: "test_cte",
      ctequery: t.nodes.selectStmt({
        targetList: [t.nodes.resTarget({
          val: t.nodes.columnRef({
            fields: [t.nodes.string({
              sval: "id"
            })]
          })
        })],
        limitOption: "LIMIT_OPTION_DEFAULT"
      })
    })],
    recursive: false
  }),
  op: "SETOP_NONE"
})"
`;

exports[`Complex AST â€” Advanced SQL with CTEs, Window Functions, Joins, and Subqueries 1`] = `
"ast.selectStmt({
  targetList: [ast.resTarget({
    name: "customer_name",
    val: ast.columnRef({
      fields: [ast.string({
        sval: "c"
      }), ast.string({
        sval: "name"
      })]
    })
  }), ast.resTarget({
    name: "total_sales",
    val: ast.columnRef({
      fields: [ast.string({
        sval: "cr"
      }), ast.string({
        sval: "total_sales"
      })]
    })
  }), ast.resTarget({
    name: "sales_rank",
    val: ast.columnRef({
      fields: [ast.string({
        sval: "cr"
      }), ast.string({
        sval: "sales_rank"
      })]
    })
  }), ast.resTarget({
    name: "customer_tier",
    val: ast.caseExpr({
      args: [ast.caseWhen({
        expr: ast.aExpr({
          kind: "AEXPR_OP",
          name: [ast.string({
            sval: "<="
          })],
          lexpr: ast.columnRef({
            fields: [ast.string({
              sval: "cr"
            }), ast.string({
              sval: "sales_rank"
            })]
          }),
          rexpr: ast.aConst({
            ival: ast.ival({})
          })
        }),
        result: ast.aConst({
          sval: ast.sval({
            0: "P",
            1: "r",
            2: "e",
            3: "m",
            4: "i",
            5: "u",
            6: "m"
          })
        })
      }), ast.caseWhen({
        expr: ast.aExpr({
          kind: "AEXPR_OP",
          name: [ast.string({
            sval: "<="
          })],
          lexpr: ast.columnRef({
            fields: [ast.string({
              sval: "cr"
            }), ast.string({
              sval: "sales_rank"
            })]
          }),
          rexpr: ast.aConst({
            ival: ast.ival({})
          })
        }),
        result: ast.aConst({
          sval: ast.sval({
            0: "G",
            1: "o",
            2: "l",
            3: "d"
          })
        })
      }), ast.caseWhen({
        expr: ast.aExpr({
          kind: "AEXPR_OP",
          name: [ast.string({
            sval: "<="
          })],
          lexpr: ast.columnRef({
            fields: [ast.string({
              sval: "cr"
            }), ast.string({
              sval: "sales_rank"
            })]
          }),
          rexpr: ast.aConst({
            ival: ast.ival({})
          })
        }),
        result: ast.aConst({
          sval: ast.sval({
            0: "S",
            1: "i",
            2: "l",
            3: "v",
            4: "e",
            5: "r"
          })
        })
      })],
      defresult: ast.aConst({
        sval: ast.sval({
          0: "B",
          1: "r",
          2: "o",
          3: "n",
          4: "z",
          5: "e"
        })
      })
    })
  }), ast.resTarget({
    name: "recent_order_count",
    val: ast.subLink({
      subLinkType: "EXPR_SUBLINK",
      subselect: ast.selectStmt({
        targetList: [ast.resTarget({
          val: ast.funcCall({
            funcname: [ast.string({
              sval: "count"
            })],
            args: [ast.aStar({})]
          })
        })],
        fromClause: [ast.rangeVar({
          relname: "orders",
          inh: true,
          relpersistence: "p",
          alias: ast.aliasname({
            0: "o",
            1: "2"
          })
        })],
        whereClause: ast.boolExpr({
          boolop: "AND_EXPR",
          args: [ast.aExpr({
            kind: "AEXPR_OP",
            name: [ast.string({
              sval: "="
            })],
            lexpr: ast.columnRef({
              fields: [ast.string({
                sval: "o2"
              }), ast.string({
                sval: "customer_id"
              })]
            }),
            rexpr: ast.columnRef({
              fields: [ast.string({
                sval: "c"
              }), ast.string({
                sval: "id"
              })]
            })
          }), ast.aExpr({
            kind: "AEXPR_OP",
            name: [ast.string({
              sval: ">="
            })],
            lexpr: ast.columnRef({
              fields: [ast.string({
                sval: "o2"
              }), ast.string({
                sval: "order_date"
              })]
            }),
            rexpr: ast.funcCall({
              funcname: [ast.string({
                sval: "current_date"
              })],
              args: []
            })
          })]
        }),
        limitOption: "LIMIT_OPTION_DEFAULT",
        op: "SETOP_NONE"
      })
    })
  })],
  fromClause: [ast.joinExpr({
    jointype: "JOIN_INNER",
    larg: ast.rangeVar({
      relname: "customers",
      inh: true,
      relpersistence: "p",
      alias: ast.aliasname({
        0: "c"
      })
    }),
    rarg: ast.rangeVar({
      relname: "customer_rankings",
      inh: true,
      relpersistence: "p",
      alias: ast.aliasname({
        0: "c",
        1: "r"
      })
    }),
    quals: ast.aExpr({
      kind: "AEXPR_OP",
      name: [ast.string({
        sval: "="
      })],
      lexpr: ast.columnRef({
        fields: [ast.string({
          sval: "c"
        }), ast.string({
          sval: "id"
        })]
      }),
      rexpr: ast.columnRef({
        fields: [ast.string({
          sval: "cr"
        }), ast.string({
          sval: "customer_id"
        })]
      })
    })
  })],
  whereClause: ast.boolExpr({
    boolop: "AND_EXPR",
    args: [ast.aExpr({
      kind: "AEXPR_OP",
      name: [ast.string({
        sval: ">"
      })],
      lexpr: ast.columnRef({
        fields: [ast.string({
          sval: "cr"
        }), ast.string({
          sval: "total_sales"
        })]
      }),
      rexpr: ast.aConst({
        ival: ast.ival({})
      })
    }), ast.aExpr({
      kind: "AEXPR_OP",
      name: [ast.string({
        sval: "IS NOT"
      })],
      lexpr: ast.columnRef({
        fields: [ast.string({
          sval: "c"
        }), ast.string({
          sval: "status"
        })]
      }),
      rexpr: ast.aConst({
        sval: ast.sval({
          0: "i",
          1: "n",
          2: "a",
          3: "c",
          4: "t",
          5: "i",
          6: "v",
          7: "e"
        })
      })
    })]
  }),
  sortClause: [ast.sortBy({
    node: ast.columnRef({
      fields: [ast.string({
        sval: "cr"
      }), ast.string({
        sval: "sales_rank"
      })]
    }),
    sortby_dir: "SORTBY_ASC",
    sortby_nulls: "SORTBY_NULLS_DEFAULT"
  })],
  limitCount: ast.aConst({
    ival: ast.ival({})
  }),
  limitOption: "LIMIT_OPTION_COUNT",
  withClause: ast.ctes({
    0: ast.commonTableExpr({
      ctename: "sales_summary",
      ctequery: ast.selectStmt({
        targetList: [ast.resTarget({
          name: "customer_id",
          val: ast.columnRef({
            fields: [ast.string({
              sval: "customer_id"
            })]
          })
        }), ast.resTarget({
          name: "total_sales",
          val: ast.funcCall({
            funcname: [ast.string({
              sval: "sum"
            })],
            args: [ast.columnRef({
              fields: [ast.string({
                sval: "amount"
              })]
            })]
          })
        }), ast.resTarget({
          name: "avg_order_value",
          val: ast.funcCall({
            funcname: [ast.string({
              sval: "avg"
            })],
            args: [ast.columnRef({
              fields: [ast.string({
                sval: "amount"
              })]
            })]
          })
        })],
        fromClause: [ast.rangeVar({
          relname: "orders",
          inh: true,
          relpersistence: "p"
        })],
        whereClause: ast.aExpr({
          kind: "AEXPR_OP",
          name: [ast.string({
            sval: ">="
          })],
          lexpr: ast.columnRef({
            fields: [ast.string({
              sval: "order_date"
            })]
          }),
          rexpr: ast.aConst({
            sval: ast.sval({
              0: "2",
              1: "0",
              2: "2",
              3: "3",
              4: "-",
              5: "0",
              6: "1",
              7: "-",
              8: "0",
              9: "1"
            })
          })
        }),
        groupClause: [ast.columnRef({
          fields: [ast.string({
            sval: "customer_id"
          })]
        })],
        limitOption: "LIMIT_OPTION_DEFAULT",
        op: "SETOP_NONE"
      })
    }),
    1: ast.commonTableExpr({
      ctename: "customer_rankings",
      ctequery: ast.selectStmt({
        targetList: [ast.resTarget({
          name: "customer_id",
          val: ast.columnRef({
            fields: [ast.string({
              sval: "customer_id"
            })]
          })
        }), ast.resTarget({
          name: "total_sales",
          val: ast.columnRef({
            fields: [ast.string({
              sval: "total_sales"
            })]
          })
        }), ast.resTarget({
          name: "sales_rank",
          val: ast.windowFunc({
            winfnoid: 3133,
            wintype: 20,
            args: [],
            winref: 1,
            winstar: false,
            winagg: false
          })
        }), ast.resTarget({
          name: "sales_percentile",
          val: ast.windowFunc({
            winfnoid: 3974,
            wintype: 701,
            args: [],
            winref: 2,
            winstar: false,
            winagg: false
          })
        })],
        fromClause: [ast.rangeVar({
          relname: "sales_summary",
          inh: true,
          relpersistence: "p"
        })],
        windowClause: [ast.windowDef({
          name: "sales_window",
          orderClause: [ast.sortBy({
            node: ast.columnRef({
              fields: [ast.string({
                sval: "total_sales"
              })]
            }),
            sortby_dir: "SORTBY_DESC",
            sortby_nulls: "SORTBY_NULLS_DEFAULT"
          })]
        })],
        limitOption: "LIMIT_OPTION_DEFAULT",
        op: "SETOP_NONE"
      })
    })
  }),
  op: "SETOP_NONE"
})"
`;

exports[`Enhanced AST generation with runtime schema â€” wrapped vs unwrapped nodes 1`] = `
"t.nodes.selectStmt({
  targetList: [t.nodes.resTarget({
    val: t.nodes.columnRef({
      fields: [t.nodes.aStar({})]
    })
  })],
  fromClause: [t.nodes.rangeVar({
    relname: "test_table",
    inh: true,
    relpersistence: "p"
  })],
  limitOption: "LIMIT_OPTION_DEFAULT",
  op: "SETOP_NONE"
})"
`;
